File: backend/app/schemas/stack.py
"""Stack JSON schema with hostability constraints."""
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, field_validator, model_validator


class StackJson(BaseModel):
    frontend_framework: Literal["nextjs", "react", "vue", "svelte"] = "nextjs"
    backend_framework: Literal["fastapi", "express", "nestjs", "django"] = "fastapi"
    database: Literal["postgresql", "mysql", "sqlite"] = "postgresql"
    database_provider: Literal["supabase", "neon", "planetscale", "local"] = "supabase"
    hosting_frontend: Literal["vercel"] = "vercel"
    hosting_backend: Literal["supabase", "vercel"] = "supabase"

    @field_validator("hosting_frontend")
    @classmethod
    def frontend_must_be_vercel(cls, v: str) -> str:
        if v != "vercel":
            raise ValueError("Frontend hosting is locked to 'vercel' in MVP")
        return v

    @field_validator("hosting_backend")
    @classmethod
    def backend_hosting_constraint(cls, v: str) -> str:
        if v not in ("supabase", "vercel"):
            raise ValueError("Backend hosting must be 'supabase' or 'vercel'")
        return v

    @model_validator(mode="after")
    def validate_hostability(self) -> StackJson:
        if self.database_provider == "local":
            raise ValueError("Local database providers are not hostable in MVP")

        if self.database_provider == "supabase" and self.database != "postgresql":
            raise ValueError("Supabase provider requires PostgreSQL")
        if self.database_provider == "neon" and self.database != "postgresql":
            raise ValueError("Neon provider requires PostgreSQL")
        if self.database_provider == "planetscale" and self.database != "mysql":
            raise ValueError("PlanetScale provider requires MySQL")

        if self.database == "sqlite":
            raise ValueError("SQLite is not supported for hosted MVP deployments")

        if self.hosting_backend == "supabase" and self.database_provider != "supabase":
            raise ValueError("Supabase backend hosting requires Supabase database provider")

        return self


DEFAULT_STACK = StackJson()

File: frontend/src/types/stack.ts
export interface StackJson {
  frontend_framework: "nextjs";
  backend_framework: "nextjs";
  database: "none";
  database_provider: "none";
  hosting_frontend: "vercel";
  hosting_backend: "vercel";
}

export const DEFAULT_STACK: StackJson = {
  frontend_framework: "nextjs",
  backend_framework: "nextjs",
  database: "none",
  database_provider: "none",
  hosting_frontend: "vercel",
  hosting_backend: "vercel",
};

export const FIXED_STACK_LABELS: Array<{ key: keyof StackJson; label: string }> = [
  { key: "frontend_framework", label: "Frontend" },
  { key: "backend_framework", label: "Backend" },
  { key: "database", label: "Database" },
  { key: "database_provider", label: "Database Provider" },
  { key: "hosting_frontend", label: "Frontend Hosting" },
  { key: "hosting_backend", label: "Backend Hosting" },
];

export function validateStackHostability(stack: StackJson): string | null {
  const mismatches = Object.entries(DEFAULT_STACK).filter(
    ([key, expected]) => stack[key as keyof StackJson] !== expected
  );
  if (mismatches.length > 0) {
    return "Stack profile is fixed to Next.js full-stack on Vercel";
  }
  return null;
}

